### ISSUE TITLE
Best way to persist whitespace when dragging and dropping?

### ISSUE THREAD

**Issue by chrisbreiding** [metafizzy/packery#337]
First off, thank you very much for this library. It's awesome.

My use case for Packery is having a dashboard of widgets that users can drag and drop to re-arrange their order and layout. The order is persisted to a database. This works great for the most part, but there are cases where the layout changes after refreshing because the whitespace isn't preserved. For example, given the following: 

![original layout](https://cloud.githubusercontent.com/assets/1157043/13223168/2578c29c-d951-11e5-9f80-aa8187a4ce6b.png)

If a user moves widget 'c' to the right, like so:

![after dragging c to right](https://cloud.githubusercontent.com/assets/1157043/13223190/397eae28-d951-11e5-8158-2054f26a1f55.png)

and then refreshes, widget 'c' ends up back on the left, because all that's persisted is the order and that doesn't actually change between the original layout and dragging 'c' to the right. 

[Here is a reduced test case demonstrating this](http://codepen.io/chrisbreiding/pen/LGKgbM). Note that I don't expect the code in this test case to preserve the white space, but am wondering if there _is_ a way to persist the whitespace so that the widgets display in the same layout upon refreshing in which they are left after dragging and dropping. 

I think that #179 might help with this, using `getItemElements` (or perhaps a new API that includes the dummy elements) to persist the order of the widgets interspersed with dummy elements. Or if there is a way to do this currently, please advise. Thank you!

**Comment by desandro** [comment]
Thanks for reporting this issue.

Saving and restoring dragged item positions can be complicated, but it's doable.  See [Save & restore drag position demo](http://codepen.io/desandro/pen/PZrXVv), or simpler [save & restore demo that uses fixed pixel sizes](http://codepen.io/desandro/pen/wMLNyz)

This demo saves item positions according to an HTML attribute, in this case `data-item-id`

``` html
<div class="grid-item" data-item-id="1"></div>
<div class="grid-item" data-item-id="2"></div>
<div class="grid-item" data-item-id="3"></div>
```

Two custom methods are added to Packery, `getShiftPositions` and `initShiftLayout`.

``` js
// get JSON-friendly data for items positions
Packery.prototype.getShiftPositions = function( attrName ) {
  attrName = attrName || 'id';
  var _this = this;
  return this.items.map( function( item ) {
    return {
      attr: item.element.getAttribute( attrName ),
      x: item.rect.x / _this.packer.width
    }
  });
};

Packery.prototype.initShiftLayout = function( positions, attr ) {
  if ( !positions ) {
    // if no initial positions, run packery layout
    this.layout();
    return;
  }
  // parse string to JSON
  if ( typeof positions == 'string' ) {
    try {
      positions = JSON.parse( positions );
    } catch( error ) {
      console.error( 'JSON parse error: ' + error );
      this.layout();
      return;
    }
  }
  
  attr = attr || 'id'; // default to id attribute
  this._resetLayout();
  // set item order and horizontal position from saved positions
  this.items = positions.map( function( itemPosition ) {
    var selector = '[' + attr + '="' + itemPosition.attr  + '"]'
    var itemElem = this.element.querySelector( selector );
    var item = this.getItem( itemElem );
    item.rect.x = parseFloat( itemPosition.x ) * this.packer.width;
    return item;
  }, this );
  this.shiftLayout();
};
```

Initialize Packery, but [disable `initLayout`](http://packery.metafizzy.co/options.html#initlayout). Then get saved position data, in this case from `localStorage`. `initShiftLayout` requires position object and the HTML attribute (`data-item-id`).

``` js
// init Packery
var $grid = $('.grid').packery({
  itemSelector: '.grid-item',
  columnWidth: '.grid-sizer',
  percentPosition: true,
  initLayout: false // disable initial layout
});

// get saved dragged positions
var initPositions = localStorage.getItem('dragPositions');
// init layout with saved positions
$grid.packery( 'initShiftLayout', initPositions, 'data-item-id' );
```

Make items draggable. On [`dragItemPositioned`](http://packery.metafizzy.co/draggable.html#dragitempositioned), save item position to localStorage. The custom `getShiftPositions` requires that HTML attribute.

``` js
// make draggable
$grid.find('.grid-item').each( function( i, itemElem ) {
  var draggie = new Draggabilly( itemElem );
  $grid.packery( 'bindDraggabillyEvents', draggie );
});

// save drag positions on event
$grid.on( 'dragItemPositioned', function() {
  // save drag positions
  var positions = $grid.packery( 'getShiftPositions', 'data-item-id' );
  localStorage.setItem( 'dragPositions', JSON.stringify( positions ) );
});
```

It works, but it's not 100% operational with the rest of Packery:

1. Having different layouts is not supported. You can't have 2 columns for mobile and 4 columns for desktop.
2. Adding items via server is not supported.
3. Horizontal layouts and right-to-left layouts are not supported.

**+1 this issue** if you'd like to see this functionality properly get added to Packery.

**Comment by iliyaZelenko** [comment]
+1

**Comment by GarrisonJ** [comment]
+1

**Comment by DannyR83** [comment]
+1

**Comment by pliablepixels** [comment]
+1 - also @desandro would it be possible to post above examples with vanilla JS?

**Comment by pliablepixels** [comment]
Works great with angular JS - look forward to resolving my other issue about sizing each element and I'm all set :) thank you!

**Comment by maclof** [comment]
+1 Love I would love to see this!

**Comment by mihanshilov** [comment]
+1

**Comment by sameerpanjwani** [comment]
+1

**Comment by creisman** [comment]
+1

**Comment by coolwebs** [comment]
+1

**Comment by pliablepixels** [comment]
For those adding +1 here, I am already using the code that @desandro posted above - it works very well for me. Is something missing in it?

**Comment by desandro** [comment]
@pliablepixels It should work with the demo provided, but there many edge cases it may not be able to handle: i.e. if there's a discrepancy between the current items and the saved position data to start with.

**Comment by pliablepixels** [comment]
yup, thats true, I am handling that at app level and am using your code for production, but would love to see this as a formal feature 👍

**Comment by alvarotrigo** [comment]
+1

**Comment by umbertoo** [comment]
+1

**Comment by mbeaudru** [comment]
+1

**Comment by afreeland** [comment]
@desandro Thanks for the code snippets, helped me get persistent dashboards much quicker =) ... look forward to state persistence feature at some point.

**Comment by elira** [comment]
+1

**Comment by iciodesign** [comment]
+1
Adding items via server would be amazing!

**Comment by jeffaltman** [comment]
+1
I am storing the order in a database but would really like to store and retrieve the layout (including white space).

**Comment by AliceIW** [comment]
+1

**Comment by smeyerSL** [comment]
How I solved this: I'm storing the top and left  CSS attributes to the database. When loading my dashboard I add a data-top and data-left atribute to my div which is representing a grid element.
Then when initializing packery and going trough all grid elements to make it draggable with draggabilly I use packeries fit method together with the top and left value to position each grid element.

`        
$grid.find('.grid-item').each( function( i, gridItem ) {
  var top =  $(gridItem).attr('data-top'),
        left = $(gridItem).attr('data-left'),
        draggie = new Draggabilly( gridItem );
        $grid.packery('bindDraggabillyEvents', draggie);
        $grid.packery('fit', gridItem, left, top);
});
`

**Comment by alvarotrigo** [comment]
@smeyerSL that won't position them properly when loading them in a different resolution. Which is quite common if you use multiple screens.

**Comment by smeyerSL** [comment]
It's about persisting the exact position of every grid-item relative to the grid container they are in. When you change the screen resolution then you can still allow to reposition the items. But at first I want to persist and restore a specific state. And thats a quite simple way to achieve that in a quick, maybe basic manner.

**Comment by ashubham** [comment]
So this seems to fail, when there is a gutter with %age position. http://codepen.io/anon/pen/XKLWJY

Try, make the screen small, drag position and then make the screen bigger and refresh, In some configurations one can find that the positions are now staggered. (If only pythagoras were alive :( )

**Comment by desandro** [comment]
From @ashubham 

---

So the solution mentioned in #337 seems to fail, when there is a gutter with %age position. http://codepen.io/anon/pen/XKLWJY

Try, make the screen small, drag position and then make the screen bigger and refresh, In some configurations one can find that the positions are now staggered.

The percentage position of the items is not maintained with screen size. There are slight rounding errors like with a gutter of 1%, two column 49.5% items should be, left: 0% and left:50.5% respectively. But its like left:0% and left:50.348% or something similar depending on the screensize.

**Comment by Firestorm-Graphics** [comment]
could this be adapted to save the positions to database rather than local storage? looking at something a little like wordpress behaviour

**Comment by desandro** [comment]
@Firestorm-Graphics Yes. The demo uses Local storage as an example. So long as you can format the data in the same manner it should work.

**Comment by markwebsterUK** [comment]
For me this isn't saving the vertical ordering, and adding adding item.rect.y to the data that gets saved (and then positioning on the y axis on load) doesn't seem to have an effect. the "column" each item is in is restored, but the vertical ordering is not.

**Comment by pliablepixels** [comment]
@markwebsterUK - hmm I don't seem to have your problem. [Here is a video ](https://youtu.be/Z2brZff-bgo)- I'm rearranging both size and order and it works fine. I'm facing other issues with imagesLoading firing early, but not this.

**Comment by markwebsterUK** [comment]
@pliablepixels yeah, realised it was me messing with the function to try and account for new/missing blocks that broke the ordering. I deleted the comment, but apparently you're too quick off the mark!

**Comment by markwebsterUK** [comment]
Quick and dirty, but I've added a couple of bits that will filter out missing items, which used to cause Packery to fail, and adds any new items. It's not all that efficient at the moment (loop within a loop), but it seems to work.
```javascript
Packery.prototype.initShiftLayout = function( positions, attr ) {
    if ( !positions ) {
        // if no initial positions, run packery layout
        this.layout();
        return;
    }
    // parse string to JSON
    if ( typeof positions == 'string' ) {
        try {
            positions = JSON.parse( positions );
        } catch( error ) {
            notification.alert(str.get_string('error', 'core'), 'JSON parse error: ' + error );
            this.layout();
            return;
        }
    }

    attr = attr || 'id'; // default to id attribute
    this._resetLayout();
    // set item order and horizontal position from saved positions
    this.items = positions.map( function( itemPosition ) {
        var selector = '[' + attr + '="' + itemPosition.attr  + '"]';
        var itemElem = this.element.querySelector( selector );
        var item = this.getItem( itemElem );
        if (item) {
            item.rect.x = itemPosition.x * this.packer.width;
            return item;
        }
    }, this );

/************** new bit starts here **************/

    // filter out any items that no longer exist
    this.items = this.items.filter(function(item) {
        return item !== undefined;
    });
    // add new items
    var newitems = [];
    var currentitems = this.items;
    $(this.$element[0]).find(this.options.itemSelector).each(function(i, e) {
        var itemexists = false;
        for (var x = 0; x < currentitems.length; x++) {
            if (currentitems[x].element == e) {
                itemexists = true;
                break;
            }
        }
        if (!itemexists) {
            newitems.push(e);
        }
    });
    this.addItems(newitems);

    this.shiftLayout();
};
```

**Comment by joecacti** [comment]
@pliablepixels - I'm trying to get this to work in my angular app.  Care to share a codepen on how you managed to achieve this?

**Comment by pliablepixels** [comment]
@joecacti - you mean initShiftLayout ? I stuffed in the function to packery's JS file exactly as @desandro wrote above and am directly calling it via `pckry.initShiftLayout()`.  Code starts [here](https://github.com/pliablepixels/zmNinja/blob/master/www/js/MontageCtrl.js#L170). I didn't write a directive wrapping it - I wrapped it in [$timeout instead](http://stackoverflow.com/a/23073968) to make sure angular calls its digest cycle safely.

**Comment by pliablepixels** [comment]
@ashubham did you solve the staggering issue?

**Comment by digelim** [comment]
@desandro how the same solution will look like using vanilla?

**Comment by desandro** [comment]
@digelim [See demo on CodePen](http://codepen.io/desandro/pen/WovjXq). Vanilla JS init code would be:

``` js
// init Packery
var grid = document.querySelector('.grid');
var pckry = new Packery( grid, {
  itemSelector: '.grid-item',
  columnWidth: '.grid-sizer',
  percentPosition: true,
  initLayout: false, // disable initial layout
});

// get saved dragged positions
var initPositions = localStorage.getItem('dragPositions');
// init layout with saved positions
pckry.initShiftLayout( initPositions, 'data-item-id' );

// make draggable
var items = grid.querySelectorAll('.grid-item');
for ( var i=0; i < items.length; i++ ) {
  var itemElem = items[i];
  var draggie = new Draggabilly( itemElem );
  pckry.bindDraggabillyEvents( draggie );
}

// save drag positions on event
pckry.on( 'dragItemPositioned', function() {
  var positions = pckry.getShiftPositions('data-item-id');
  // save drag positions
  localStorage.setItem( 'dragPositions', JSON.stringify( positions ) );
});
```

**Comment by pliablepixels** [comment]
Hello @desandro - I know this is beta, but hoping for some thoughts - I am facing problems where the images re-locate when I resize the window.
Here is an example. When the app first starts, I use `pckry.initShiftLayout(positions,'data-item-id')` and it loads up perfectly. The moment I resize the window, the images re-arrange. 

I've tried to trap onrpsize and do a `pckry.initShiftLayout(..)` in it , but it doesn't help - the images still move. How do I stop it from moving around when there is sufficient place for the images to stay as is ?

![dec-04-2016 18-56-43](https://cloud.githubusercontent.com/assets/4116654/20870309/cbbf7b1c-ba53-11e6-888c-03bfeededc0d.gif)

**Comment by desandro** [comment]
@pliablepixels There's an undocumented option `shiftPercentResize` that should address this. Try setting `shiftPercentResize: true`.

**Comment by pliablepixels** [comment]
@desandro - perfect! I think I finally have this worked out where it works 100% of the time on Desktops/Android and iOS via angular. 🙏

**Comment by pliablepixels** [comment]
@desandro - correction, there is still one last issue I'm facing. It was mentioned earlier by someone else and I don't have a solution. Given we are on a roll on solving issues, I'll throw this out out.

**Problem description:** When I save a profile that does not have equally sized images, I almost always see staggering.
**Problem resolution:** A 100% fix is simply to resize the window (even by a couple of pixels!)
**Trying to hack it:** Given a resize always fixes, I tried calling `pckry.onresize()` forcefully after all images are laid out. No dice. I tried calling `pckry.shifyLayout()` forcefully after all images are laid out. No dice. I tried calling `pckry.layout()` after all images are laid out. That works, but is not desirable as the images re-orient.
So given a physical resize of the window fixes stuff, how do I emulate the same behavior if that is what it takes to hack it?

Demo:
![stagger2](https://cloud.githubusercontent.com/assets/4116654/20909042/2b3a1e16-bb27-11e6-9085-8de441b4c5e3.gif)

**Comment by creisman** [comment]
It's been a long time since I worked with this, but I think I had this problem and it ended up being a problem with rounding. I'm looking at the code I ended up writing and the only substantial difference I can find is that instead of 'this.packer.width' I used 'item.rect.width'. But I don't remember if that was to solve this problem or if that was just because I added margin/padding to something...

I also call this.shiftLayout(); AFTER this._resetLayout() but before positions.map. I don't remember if that was necessary to establish the widths or if that was simply because I forgot to clean it up properly.

Hopefully one (or both) of these helps!

**Comment by buggedcom** [comment]
Since not all of the items may exist in the stored positions it is necessary to filter out any items that do not exist.

```js
                // set item order and horizontal position from saved positions
                this.items = positions.map(
                    function( itemPosition ) {
                        var selector = '[' + attr + '="' + itemPosition.attr  + '"]';
                        var itemElem = this.element.querySelector( selector );
                        var item = this.getItem( itemElem );
                        if(!item) {
                            return false;
                        }
                        
                        item.rect.x = Math.round(itemPosition.x * this.packer.width);
                        return item;
                    },
                    this
                ).filter(
                    function(value) {
                        return value !== false;
                    }
                );
```

**Comment by keyscores** [comment]
+1 would like to see this as a built-in method.

**Comment by shaani88** [comment]
Hi Desandro, Could you please try to show an example where we can add new grid items to the grid and save them to localstorage and delete them too? I have been trying for days but could not get the newly added items to save in localstorage. thanks a lot

**Comment by ERPG** [comment]
Hey Desandro, thank you for the Demo, it work very well!.. remember guys if it doesn't work, go to local Storage, clear and erase all, and then try again...
Regards

**Comment by CTKRocks** [comment]
How would I achieve this with a horizontal layout?

**Comment by CTKRocks** [comment]
Think I got it mostly working:
```javascript
Packery.prototype.getShiftPositions = function( attrName ) {
  attrName = attrName || 'id';
  var _this = this;
  return this.items.map( function( item ) {
    return {
      attr: item.element.getAttribute( attrName ),
      x: item.rect.x / _this.packer.width,
      y: item.rect.y / _this.packer.height
    }
  });
};

Packery.prototype.initShiftLayout = function( positions, attr ) {
    if ( !positions ) {
        // if no initial positions, run packery layout
        this.layout();
        return;
    }
    // parse string to JSON
    if ( typeof positions == 'string' ) {
        try {
            positions = JSON.parse( positions );
        } catch( error ) {
            console.error( 'JSON parse error: ' + error );
            this.layout();
            return;
        }
    }

    attr = attr || 'id'; // default to id attribute
    this._resetLayout();
    // set item order and horizontal position from saved positions
    this.items = positions.map( function( itemPosition ) {
        var selector = '[' + attr + '="' + itemPosition.attr  + '"]';
        var itemElem = this.element.querySelector( selector );
        var item = this.getItem( itemElem );
        if (item) {
            item.rect.x = itemPosition.x * this.packer.width;
            item.rect.y = itemPosition.y * this.packer.height;
            return item;
        }
    }, this );

    // filter out any items that no longer exist
    this.items = this.items.filter(function(item) {
        return item !== undefined;
    });
    // add new items
    var newitems = [];
    var p = this;
    $(this.$element[0]).find(this.options.itemSelector).each(function(i, e) {
        if (!p.getItem(e)) {
            newitems.push(e);
        }
    });
    this.addItems(newitems);

    this.shiftLayout();
};
```

**Comment by chanaka1** [comment]
any idea why i get "_resetLayout 'columnWidth' of undefined" when called "  this._resetLayout();" from Jquery

```
 var $container = $('.dashboard');
                $container.packery({
                    columnWidth: '.grid-sizer',
                    itemSelector: '.grid-item',
                    percentPosition: true,
                    initLayout: false
                });
var packery = $('.dashboard').data('packery');
packery.__proto__._resetLayout();

// After : set item order and horizontal position from saved positions etc..
```

Trying to do this from JQuery https://codepen.io/anon/pen/KXqVpx?editors=0010

**Comment by Tezar** [comment]
@CTKRocks thanks for snippet. 

Any idea why would packer.height be infinity?

**Comment by SirZdanius** [comment]
+1 to see this as a full feature!

Kudos on the awesome library and this demo works fine but I will definitely need this on mobile, from server **and** for a horizontal layout for my particular use case.

**Comment by NikolaR** [comment]
+1

**Comment by 23creative** [comment]
@desandro Firstly thank you the awesome library and everyone else for the additional code

After a lot of console logging I have got pretty close to removing the staggering issue. It does take a little testing to get it perfect though. When using responsive layout I found that `setPosition()` was using `parseInt()` which rounds the x value. When working with the calculation for percentages in my layout it was changing the x value of an element to 33.234% as opposed to 33.333%

so I added 

```
Packery.Item.prototype.setPosition = function( x, y ) {
        if(x > 0){
            //the +0.07 is my tweaking value to get the final calculation perfect
            this.position.x = parseFloat( x )+0.07;
        }else{
            this.position.x = parseFloat( x );
        }
        this.position.y = parseFloat( y );
    };
```

**Comment by Ilirwork** [comment]
@desandro 
Hi
I've several days that I'm trying to make that the example 'Packery - save & restore drag position'  can work with resizable. Resizable in that example works only in horizontal, I think due to 'padding-bottom'.
 Please can you explain how I can make possible that 'resizable' works?
What I need is a probject packery where Items are append, positioned and sized both Not in pixel but in Percentage. 
Thanks a lot

**Comment by svet007** [comment]
👍

**Comment by qJake** [comment]
This would be a great feature to have - I have a use case where I want the user to be able to use Packery+Draggabilly to set the exact position of elements and have it retain that same position (for a user-customizable dashboard of sorts).

Calling `layout()` breaks this, so in essence what I need (and what I think is outlined in pieces above here) is the ability to initialize Packery (with drag/drop events) without it performing an auto-layout and moving things around.

**Comment by pliablepixels** [comment]
@qJake I'm not sure it requires a server side solution, does it? I store state in local storage and it works fine to restore. And yes, I use `.shiftLayout()` instead of `.layout()`

**Comment by qJake** [comment]
@pliablepixels Yes, it absolutely does not *require* something server-side, and certainly local storage is a viable option if you are designing a client-side application.

However, the application I'm designing has an ASP.NET Core backend, so being able to send the dashboard layout to the server, and then have the server re-create that same layout (while still being able to initialize and use Packery and Draggabilly functionality) was necessary in my case. Hence, reading the X/Y coordinates from the actual item's style (which the server renders in the view) was how I got my solution to work. 🙂

**Comment by prajapati-gautam** [comment]
if i keep space upside of element then it will not restoring it back.
you can see here [demo](https://codepen.io/desandro/pen/PZrXVv)

**Comment by qJake** [comment]
@prajapati-gautam You'd be better off submitting a new issue, your issue is not related to this one.

And I can tell you, it's not possible to have whitespace above an element - all elements will attempt to collapse and shift to the upper-left corner of the container (or wherever your origin is). I worked around this by creating an empty element that you can see while dragging, but that does not show up otherwise. This allows for some "padding" or "blanks" that you can place wherever you want to offset elements.

**Comment by Hulk93** [comment]
can someone please help me with this.

save pre-dragged position on button click.

thanks in advance.

**Comment by aarondeloach** [comment]
I ended up writing a cookie with the container id and the sort order. I catch the cookie on the next request and save the data to the user table and set a session value with same. When I load the widgets I simply sort them against the known state.
```javascript
<script>

    var $grid = $('.grid').packery({
        itemSelector: '.grid-item',
        columnWidth: '.grid-item',
        percentPosition: true,
    });

    $grid.find('.grid-item').each( function( i, gridItem ) {
        var draggie = new Draggabilly( gridItem );
        // bind drag events to Packery
        $grid.packery( 'bindDraggabillyEvents', draggie );
    });

    function orderItems() {
        var itemElems = $grid.packery('getItemElements');
        var s = '';
        $( itemElems ).each( function( i, itemElem ) {
            s+=itemElem.id + '|'+ (i+1)+',';
        });
        document.cookie = "_dbw=" + s + "; path=/";
    }

    $grid.on( 'layoutComplete', orderItems );
    $grid.on( 'dragItemPositioned', orderItems );

</script>

```
